consensus/src/experimental/commit_phase.rs

```diff
@@ -408,49 +407,38 @@ impl CommitPhase {
                             }
                         };
-                    }
-                    retry_cv = self.timeout_event_rx.select_next_some() => {
+                }
+                retry_cv = self.timeout_event_rx.select_next_some(), if self.blocks.is_some() => {
                         if let Some(ref pending_blocks) = self.blocks {
                             if pending_blocks.block_info() == retry_cv.commit_info() {
                                 // retry broadcasting the message if the blocks are still pending
                                 tokio::spawn(broadcast_commit_vote_with_retry(self.network_sender.clone(), retry_cv, self.timeout_event_tx.clone()));
                             }
                         }
-                    }
-                    // callback event might come when self.blocks is not empty
-                    reset_event_callback = self.reset_event_rx.select_next_some() => {
+                }
+                // callback event might come when self.blocks is not empty
+                reset_event_callback = self.reset_event_rx.select_next_some() => {
                         self.process_reset_event(reset_event_callback).await.map_err(|e| ExecutionError::InternalError {
                             error: e.to_string(),
                         })
                         .unwrap();
-                    }
-                    complete => break,
                 }
-                report_err!(
-                    // check if the blocks are ready to commit
-                    self.check_commit().await,
-                    "Error in checking whether self.block is ready to commit."
-                );
-            }
-
-            select! {
-                CommitChannelType(blocks, ordered_ledger_info, callback) = self.commit_channel_recv.select_next_some() => {
+                CommitChannelType(blocks, ordered_ledger_info, callback) = self.commit_channel_recv.select_next_some(),
+                    if self.blocks.is_none() => {
                     report_err!(
                         // receive new blocks from execution phase
                         self.process_executed_blocks(blocks, ordered_ledger_info, callback)
                             .await,
                         "Error in processing received blocks"
                     );
                 }
-                // callback event might come when self.blocks is none
-                reset_event_callback = self.reset_event_rx.select_next_some() => {
-                    self.process_reset_event(reset_event_callback).await.map_err(|e| ExecutionError::InternalError {
-                        error: e.to_string(),
-                    })
-                    .unwrap();
-                }
-                complete => break,
+                else => break,
             }
+            report_err!(
+                // check if the blocks are ready to commit
+                self.check_commit().await,
```

Comment:
this can be moved to message branch? we probably can have a wrapper function process_commit_message

Reply:
good idea